`ranges::distance` fails to emulate the Standard's two overload presentation

CaseyCarter opened the pull request:
By not properly constraining itself with the union of the constraints of those two overloads. If we weren't trying to do something fancy just to avoid a bit of non-subsumption concept overloading, we wouldn't have written this bug in the first place. Let's split out into two overloads.

Test coverage is incoming in the LLVM update, against which I've verified the fix.

Merge conflict with #3024

strega-nil-ms commented:
Is there a reason not to just do the two-overload version, given that the two-overload version is how the standard is written?

CaseyCarter commented:
> Is there a reason not to just do the two-overload version, given that the two-overload version is how the standard is written?

There are exactly two reasons why an STL implementation would take advantage of as-if to do something differently from how it's phrased in the Standard: performance and throughput. This is the latter.

We generally prefer `if constexpr` dispatch to concept overloading because it's quicker, even in cases like this where subsumption isn't a factor (normalization and subsumption checking are the really expensive bits). That said, I could be convinced that this bug wouldn't have happened if I hadn't tried to be fancy, and that I should fix the wording in LWG-3664 if I want to do something different.

strega-nil-ms commented:
I would agree with the latter point. If they took the same parameters, it would be different IMO, but doing this weird `decay_t` stuff to allow taking _either_ a `const I&` _OR_ an `I` by value is a lot.

CaseyCarter commented:
> I would agree with the latter point. If they took the same parameters, it would be different IMO, but doing this weird `decay_t` stuff to allow taking _either_ a `const I&` _OR_ an `I` by value is a lot.

Split into two overloads, reverified against the incoming libc++ test, and rewrote the commit message.

strega-nil-ms commented on stl/inc/xutility:
```diff
@@ -2507,16 +2507,19 @@ namespace ranges {
+        template <class _It, sentinel_for<_It> _Se>
+            requires(!sized_sentinel_for<_Se, _It>)
+        _NODISCARD constexpr iter_difference_t<_It> operator()(_It _First, _Se _Last) const
+            noexcept(noexcept(_Distance_unchecked(
```
Until `_Get_unwrapped` is `noexcept`, this will not be noexcept, and MSVC with `/permissive` thinks `_Get_unwrapped` is noexcept for some godforsaken reason (this is why the test is failing)

strega-nil-ms replied:
(this should probably be an issue - making `_Get_unwrapped` noexcept)

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -2576,16 +2576,19 @@ namespace ranges {
-                _Adl_verify_range(_First, _Last);
-                return _Distance_unchecked(_Get_unwrapped(_STD move(_First)), _Get_unwrapped(_STD move(_Last)));
-            }
+        template <class _It, sentinel_for<_It> _Se>
+            requires(!sized_sentinel_for<_Se, _It>)
+        _NODISCARD constexpr iter_difference_t<_It> operator()(_It _First, _Se _Last) const
```
`// Per LWG-3664` applies to this overload too, as WG21-N4910 depicts `template<input_or_output_iterator I` but LWG-3664 has `template<class I`.

StephanTLavavej commented on tests/std/tests/P0896R4_ranges_iterator_machinery/test.cpp:
```diff
@@ -2959,7 +2959,7 @@ namespace iter_ops {
             I first{t};
             same_as<iter_difference_t<I>> auto const result = distance(move(first), default_sentinel);
-            STATIC_ASSERT(!noexcept(distance(move(first), default_sentinel))); // No conditional noexcept
+            STATIC_ASSERT(noexcept(distance(move(first), default_sentinel))); // No conditional noexcept
```
I believe that the comment here should be dropped; we are strengthening both overloads with conditional `noexcept` now (instead of just the `sized_sentinel_for<decay_t<_It>>` case).

StephanTLavavej commented:
Looks good to me. FYI @strega-nil-ms, I pushed a couple of comment changes after you approved.

CaseyCarter commented:
Thanks for finishing this up, @strega-nil-ms!

StephanTLavavej commented:
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

StephanTLavavej commented:
Thanks for going the distance here! :runner: :1st_place_medal: :joy_cat:
